src/test/java/com/google/firebase/messaging/FirebaseMessagingTest.java

```diff
@@ -0,0 +1,554 @@
+        fail("No error thrown for HTTP error");
+      } catch (ExecutionException e) {
+        assertTrue(e.getCause() instanceof FirebaseMessagingException);
+        FirebaseMessagingException error = (FirebaseMessagingException) e.getCause();
+        assertEquals("invalid-argument", error.getErrorCode());
+        assertEquals("test error", error.getMessage());
+        assertTrue(error.getCause() instanceof HttpResponseException);
+      }
+
+      assertNotNull(interceptor.getResponse());
+      HttpRequest request = interceptor.getResponse().getRequest();
+      assertEquals("POST", request.getRequestMethod());
+      assertEquals(TEST_FCM_URL, request.getUrl().toString());
+      assertEquals("Bearer test-token", request.getHeaders().getAuthorization());
+    }
+  }
+
+  @Test
+  public void testInvalidSubscribe() {
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(new MockGoogleCredentials("test-token"))
+        .setProjectId("test-project")
+        .build();
+    FirebaseApp.initializeApp(options);
+    FirebaseMessaging messaging = FirebaseMessaging.getInstance();
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    messaging.setInterceptor(interceptor);
+
+    for (TopicMgtArgs args : INVALID_TOPIC_MGT_ARGS) {
+      try {
+        messaging.subscribeToTopicAsync(args.registrationTokens, args.topic);
+        fail("No error thrown for invalid args");
+      } catch (IllegalArgumentException expected) {
+        // expected
+      }
+    }
+
+    assertNull(interceptor.getResponse());
+  }
+
+  @Test
+  public void testSubscribe() throws Exception {
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse()
+        .setContent("{\"results\": [{}, {\"error\": \"error_reason\"}]}");
+    FirebaseMessaging messaging = initMessaging(response);
+    TestResponseInterceptor interceptor = new TestResponseInterceptor();
+    messaging.setInterceptor(interceptor);
+
+    TopicManagementResponse result = messaging.subscribeToTopicAsync(
+        ImmutableList.of("id1", "id2"), "test-topic").get();
+    assertEquals(1, result.getSuccessCount());
+    assertEquals(1, result.getFailureCount());
+    assertEquals(1, result.getErrors().size());
+    assertEquals(1, result.getErrors().get(0).getIndex());
+    assertEquals("unknown-error", result.getErrors().get(0).getReason());
+
+    assertNotNull(interceptor.getResponse());
+    HttpRequest request = interceptor.getResponse().getRequest();
+    assertEquals("POST", request.getRequestMethod());
+    assertEquals(TEST_IID_SUBSCRIBE_URL, request.getUrl().toString());
+    assertEquals("Bearer test-token", request.getHeaders().getAuthorization());
+    assertEquals("true", request.getHeaders().get("access_token_auth"));
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    request.getContent().writeTo(out);
+    assertEquals("{\"to\":\"/topics/test-topic\",\"registration_tokens\":[\"id1\",\"id2\"]}",
```

Comment:
This seems overly harsh as it relies on an unspecified order during the serialization.

Reply:
Fixed by parsing into a map, and then comparing entries.